[FLINK-8966][tests] Port AvroExternalJarProgramITCase to flip6

zentol opened the pull request:
## What is the purpose of the change

This PR ports the `AvroExternalJarProgramITCase` to flip6.

## Brief change log
This test required support for setting extra jar files in the used `TestEnvironment`:

* modify `TestEnvironment#setAsContext` to pass on jarfiles/classpaths to the environment factory
* upload user-jars when submitting jobs through the MiniCluster
* allow setting extra jarfiles/classpaths on the `MiniClusterResource`

Finally:
* port `AvroExternalJarProgramITCase` to flip6

tillrohrmann commented on flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/MiniClusterResource.java:
```diff
@@ -69,6 +73,10 @@
 
+	private Collection<Path> jarFiles = Collections.emptyList();
+
+	private Collection<URL> classPaths = Collections.emptyList();
```
I think the `jarFiles` and the `classPaths` should not be part of the `MiniClusterResource`, because they belong to the client context.

zentol replied:
Yes I wasn't so sure about this either. But we either override do it like this or override part of the global state that the `MiniClusterResource` modifies when it sets the context :/

tillrohrmann replied:
Then maybe, let's not use the `MiniClusterResource` directly but instantiate a `MiniCluster` instead.

zentol replied:
that make sense

zentol replied:
We then have to duplicate the test though.

tillrohrmann replied:
I would be ok with that given that we will remove the duplicate quite soonish when we remove the legacy code.

tillrohrmann commented on flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/AvroExternalJarProgramITCase.java:
```diff
@@ -44,49 +39,24 @@
 
+	@Test
+	public void testExternalProgram() throws Exception {
+		miniClusterResource.setJarFiles(Collections.singleton(new Path(JAR_FILE)));
```
I think instead of passing the `JAR_FILE` to the `MiniClusterResource` we should set the `TestEnvironment` with the correct jar files settings explicitly in this test.

tillrohrmann replied:
Btw: Do we actually need to explicitly add the `JAR_FILE`? Doesn't the `PackagedProgram` take care of this for us?

zentol replied:
as far as i could tell, no. We actually have to set it in the environment.

tillrohrmann replied:
Ok, maybe we could verify this.

zentol replied:
The jar passed to the `PackagedProgram` is only used to create the classloader for loading the contained classes.

The jars have to be explicitly added to the JobGraph, which we here do by passing it to the environment that does it for us. In the CLI, we use [PackagedProgramUtils](https://github.com/apache/flink/blob/master/flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramUtils.java) to retrieve the jars from the PackagedProgram and register them on the jobgraph.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -606,6 +632,28 @@ public JobExecutionResult executeJobBlocking(JobGraph job) throws JobExecutionEx
 		// from the ResourceManager
 		job.setAllowQueuedScheduling(true);
 
+		CompletableFuture<Void> jarUploadFuture = currentDispatcherGateway.getBlobServerPort(rpcTimeout)
```
This looks like duplicate code.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -572,6 +576,28 @@ public void runDetached(JobGraph job) throws JobExecutionException, InterruptedE
 		// from the ResourceManager
 		job.setAllowQueuedScheduling(true);
 
+		CompletableFuture<Void> jarUploadFuture = currentDispatcherGateway.getBlobServerPort(rpcTimeout)
```
We could check whether we actually have something to upload or not.

zentol commented:
@tillrohrmann I've updated the PR.
* check whether there are jars to upload
* de-duplicate jar upload code
* reverted changes to the MiniClusterResource
* added a legacy version of the test
* updated the new version of the test to use a `MiniCluster`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -651,6 +659,34 @@ private DispatcherGateway getDispatcherGateway() throws LeaderRetrievalException
 		}
 	}
 
+	private void uploadJarFiles(final DispatcherGateway currentDispatcherGateway, final JobGraph job) throws JobExecutionException, InterruptedException {
```
Could we pass in the `JobID` and the collection of user jars instead of the `JobGraph`? We would then not set the blob keys in this function but instead return them. That way we would make this method only do the upload instead of multiple things. I think this would better follow the single responsibility principle.

zentol commented:
@tillrohrmann Could you take another look? I moved the blob upload into a separate method,

zentol commented:
merging.
